File: Top_100_Liked_Questions/494._Target_Sum/solution.py
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        # iterative dp - this works as the max sum is capped at 1000
        N = 1000
        if S > N:
            return 0
        
        mem = [[0] * (2*N+1) for _ in range(len(nums)+1)]
        mem[len(nums)][S+N] = 1
            
        for i in range(len(nums)-1, -1, -1):
            for j in range(2*N+1):
                add = mem[i+1][j - nums[i]] if j - nums[i] >= 0 else 0
                sub = mem[i+1][j + nums[i]] if j + nums[i] <= 2*N else 0
                mem[i][j] = add + sub
        
        return mem[0][N]
    
    
    def findTargetSumWays3(self, nums: List[int], S: int) -> int:
        # use index instead of subarray copy
        def helper(i, target):
            if i == len(nums):
                return 1 if target == 0 else 0
            
            return helper(i+1, target - nums[i]) + helper(i+1, target + nums[i])
        return helper(0, S)
    
    
    def findTargetSumWays2(self, nums: List[int], S: int) -> int:
        #recursive dp
        if not nums:
            return 1 if S == 0 else 0
        
        return self.findTargetSumWays(nums[1:], S-nums[0]) + self.findTargetSumWays(nums[1:], S+nums[0])

File: Top_100_Liked_Questions/494._Target_Sum/solution.java
class Solution {
    public int findTargetSumWays(int[] nums, int S) {
        // iterative dp - max sum capped at 1000
        final int max_sum = 1000;

        if (Math.abs(S) > max_sum)
            return 0;

        int[][] mem = new int[nums.length+1][max_sum*2+1];
        mem[nums.length][max_sum] = 1;

        for (int i=mem.length-2; i>-1; i--) {
            for (int j=0; j<mem[0].length; j++) {
                int add = j - nums[i] >= 0 ?  mem[i+1][j-nums[i]] : 0;
                int sub = j + nums[i] < mem[0].length ? mem[i+1][j+nums[i]] : 0;
                mem[i][j] = add + sub;
            }
        }
        return mem[0][S+max_sum];
    }
}

class Solution2 {
    public int findTargetSumWays(int[] nums, int S) {
        // recursive dp O(2^n)
        return findTargetSumWays(nums, S, 0);
    }

    public int findTargetSumWays(int[] nums, int S, int i) {
        if (i == nums.length)
            return S == 0 ? 1 : 0;

        return findTargetSumWays(nums, S - nums[i], i+1) + findTargetSumWays(nums, S + nums[i], i+1);
    }
}